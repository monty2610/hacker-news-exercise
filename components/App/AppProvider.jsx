import React, { useReducer } from 'react';
import PropTypes from 'prop-types';

export const AppContext = React.createContext(null);

const reducer = (state, action) => {
  switch (action.type) {
    case 'FETCH_POST_SUCCESS':
      return {
        posts: [...state.posts, ...action.posts],
        page: action.page,
        fetchLatest: action.fetchLatest || false,
      };
    case 'UPDATE_POST':
      return {
        ...state,
        posts: action.posts,
      };
    case 'RESET':
      return {
        ...state,
        posts: [],
        page: 1,
      };
    default:
      return state;
  }
};

const AppProvider = ({ children, posts = [] }) => {
  const [state, dispatch] = useReducer(reducer, {
    posts,
    page: 1,
    fetchLatest: false,
  });

  return (
    <AppContext.Provider value={[state, dispatch]}>
      {children}
    </AppContext.Provider>
  );
};

AppProvider.propTypes = {
  children: PropTypes.node.isRequired,
  posts: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
};

export default AppProvider;
