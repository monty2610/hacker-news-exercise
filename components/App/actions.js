export const fetchPostSuccess = (posts, page, fetchLatest) => ({
  type: 'FETCH_POST_SUCCESS',
  posts,
  page,
  fetchLatest,
});

export const updatePost = (posts) => ({
  type: 'UPDATE_POST',
  posts,
});

export const reset = () => ({
  type: 'RESET',
});
