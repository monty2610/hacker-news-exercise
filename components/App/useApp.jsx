import { useContext } from 'react';
import { AppContext } from './AppProvider';
import { fetchPostSuccess, updatePost, reset } from './actions';
import {
  getPosts,
  updatePoints,
  hidePost,
  mergePostsWithUpdate,
} from '../../utils';

export const useApp = () => {
  const [state, dispatch] = useContext(AppContext);

  const fetchNextPost = async () => {
    let { page } = state;
    const { fetchLatest } = state;
    page += 1;
    const posts = await getPosts({ page, fetchLatest });
    dispatch(fetchPostSuccess(posts, page, fetchLatest));
  };

  const fetchLatestPost = async () => {
    dispatch(reset());
    const posts = await getPosts({ fetchLatest: true, page: 1 });
    dispatch(fetchPostSuccess(posts, 1, true));
  };

  const fetchTopPost = async () => {
    dispatch(reset());
    const posts = await getPosts({ fetchLatest: false, page: 1 });
    dispatch(fetchPostSuccess(posts, 1, false));
  };

  const upvote = (objectID, points) => {
    updatePoints(objectID, points + 1);
    const posts = mergePostsWithUpdate(state.posts);
    dispatch(updatePost(posts));
  };

  const hide = (objectID) => {
    hidePost(objectID);
    const posts = mergePostsWithUpdate(state.posts);
    dispatch(updatePost(posts));
  };

  return {
    posts: state.posts,
    fetchLatest: state.fetchLatest,
    fetchNextPost,
    fetchLatestPost,
    fetchTopPost,
    upvote,
    hide,
  };
};

export default useApp;
