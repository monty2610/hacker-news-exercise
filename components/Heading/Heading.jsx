import React from 'react';
import useApp from '../App/useApp';

const Heading = () => {
  const { fetchLatest, fetchLatestPost, fetchTopPost } = useApp();

  const topClickHandler = (evt) => {
    evt.preventDefault();
    if (fetchLatest) {
      fetchTopPost();
    }
  };

  const newClickHandler = (evt) => {
    evt.preventDefault();
    if (!fetchLatest) {
      fetchLatestPost();
    }
  };

  return (
    <h2>
      <button
        type="button"
        onClick={topClickHandler}
        className={fetchLatest ? '' : 'active'}
      >
        top
      </button>
      <span> | </span>
      <button
        type="button"
        onClick={newClickHandler}
        className={fetchLatest ? 'active' : ''}
      >
        new
      </button>
      <style jsx>
        {`
          button {
            border: none;
            background: transparent;
            font-size: 12px;
          }

          .active {
            color: #fff;
          }
        `}
      </style>
    </h2>
  );
};

export default Heading;
