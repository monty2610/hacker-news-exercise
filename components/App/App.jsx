import React from 'react';
import PropTypes from 'prop-types';
import AppProvider from './AppProvider';
import Heading from '../Heading/Heading';
import PostList from '../PostList/PostList';

export const App = ({ posts }) => (
  <AppProvider posts={posts}>
    <Heading />
    <PostList />
  </AppProvider>
);

App.propTypes = {
  posts: PropTypes.shape([]).isRequired,
};

export default App;
