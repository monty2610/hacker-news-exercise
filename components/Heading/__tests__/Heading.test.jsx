import React from 'react';
import { shallow } from 'enzyme';
import Heading from '../Heading';
import * as useApp from '../../App/useApp';

describe('Heading Component', () => {
  it('Heading component should render correctly', () => {
    jest.spyOn(useApp, 'useApp').mockImplementation(() => ({}));
    const tree = shallow(<Heading />);
    expect(tree).toMatchSnapshot();
  });
});
