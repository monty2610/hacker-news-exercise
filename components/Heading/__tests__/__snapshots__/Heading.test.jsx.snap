// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading Component Heading component should render correctly 1`] = `
<nav
  className="jsx-1896995305"
>
  <img
    alt="icon"
    className="jsx-1896995305 icon"
    src="/icons/y18.gif"
  />
  <button
    className="jsx-1896995305 active"
    onClick={[Function]}
    type="button"
  >
    top
  </button>
  <span
    className="jsx-1896995305"
  >
     | 
  </span>
  <button
    className="jsx-1896995305 "
    onClick={[Function]}
    type="button"
  >
    new
  </button>
  <JSXStyle
    id="1896995305"
  >
    nav.jsx-1896995305{background:#ff6600;margin:0;padding:5px 0;}.icon.jsx-1896995305{border:1px solid #fff;vertical-align:middle;margin:0 10px;width:12px;}button.jsx-1896995305{font-size:12px;vertical-align:middle;}span.jsx-1896995305{font-size:12px;vertical-align:middle;}.active.jsx-1896995305{color:#fff;}
  </JSXStyle>
</nav>
`;
